stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1970,20 +1970,25 @@ julia> normalize(0, 1)
 NaN
 ```
 """
-function normalize(a::AbstractArray, p::Real = 2)
-    nrm = norm(a, p)
+normalize(a::AbstractArray, p::Real = 2) = _normalize(a, norm(a, p))
+
+@inline _normalize(a::AbstractArray{T}, nrm) where {T} = _normalize(promote_type(T, typeof(nrm)), a, nrm)
+Base.@constprop :aggressive @inline _normalize(a::AbstractArray{T}, nrm) where {T <: AbstractArray} = _normalize(promote_op(/, T, typeof(nrm)), a, nrm)
```

Comment:
Looks like a bad use of `promote_op`. (I know we have many of those, but that's no good excuse for adding more.)
Can't we limit the optimized version to `AbstractArray{<:Number}` and use `promote_type` there (as for the non-`AbstractArray` eltype now)?

Reply:
It's what `array / scalar` does to determine the return array type, because there doesn't seem to be any other function to take a container type and an element type and return the container type with a promoted element type.

It seemed like `promote_op` was the only alternative to actually dividing first and then taking the type of the result, the original approach. Whatever the downsides of that function, aggressive constprop seemed to eliminate them.

Why is it bad? And can you please clarify your suggestion?

Reply:
Thanks, I read that warning and I guess it wasn't sufficiently scary to understand why it is really a problem. At a minimum, it would be consistent with the comparable array-division-by-scalar operation, which is effectively what the current master is already relying on in a more roundabout way.

The fallback would skip the logic in `__normalize!` to avoid numeric instability, so that isn't a good idea.

Reply:
The fallback does not do the switch from division to multiplication with the inverse*), so the potential numeric stability issue is not introduced in the first place - maybe at the cost of performance?

*) Or does that happen somewhere is the depths of `/(A::AbstractArray, B::Number)`? If so, that would be the place to worry about the numerical instability.

Reply:
Thinking about this a bit more, wouldn't it make sense to always use `x / norm(x)` and do any potential optimizations -- w.r.t runtime and/or numerical stability -- in `/(A::AbstractArray, B::Number)`? If there's something to be gained, why should it only apply for `normalize`, but not for other divisions?

Reply:
FWIW, from cursory benchmarking, `normalize(x)` is not consistently faster than `norm(x) / x`. Wide matrices seem to favor the former, tall matrices the latter. From that, I don't think we'd loose much by just letting the `AbstractArray{<:AbstractArray}` case be handled by the fallback. I might be missing something, of course.

Reply:
By all means, reworking array/scalar division should definitely be made in a separate PR, if at all.

But restricting the `normalize(::AbtractArray, p=2)` method to `normalize(::AbtractArray{<:Number}, p=2)` and then combining with what you've done here so far but without the `_normalize(a::AbstractArray{<:AbtractArray}, nrm)` method is also a quite minimal change. And it does not introduce `promote_op` with the associated risks. It does risk performance regressions in non-`Number`-eltype cases, though it doesn't look bad at first glance. I'd prefer a slight performance regression over unsound behavior. And any performance regressions by suddenly getting an `Any` eltype would certainly be worse.

Reply:
Almost every place in LinearAlgebra that calls `copymutable_oftype` also uses `promote_op`.

And again, `promote_op` isn't being introduced here, it is already being relied on.

If you or someone else wishes to take responsibility for changing the behavior or introducing a performance regression, they can do so. I'm not going to do that.

Reply:
> Almost every place in LinearAlgebra that calls `copymutable_oftype` also uses `promote_op`.

Yes, and that's no good. Let's try to not make it worse, at least.
 
> And again, `promote_op` isn't being introduced here, it is already being relied on.

Oh, I fail to see how that is the case (for `normalize`, specifically).

Reply:
>> And again, promote_op isn't being introduced here, it is already being relied on.

> Oh, I fail to see how that is the case (for normalize, specifically).

Apologies, that seems to be wrong.

In any case, I'm done changing this and you're welcome to take over.